packages/experience/src/pages/OneTimeToken/index.tsx

```diff
@@ -81,53 +100,78 @@ const OneTimeToken = () => {
       }
     },
     [
-      preRegisterErrorHandler,
+      preSignInErrorHandler,
       asyncIdentifyUserAndSubmit,
       handleError,
       redirectTo,
-      signInWithOneTimeToken,
+      registerWithOneTimeToken,
     ]
   );
 
+  // Single effect: validate params, run terms gating once, then proceed to submission with idempotency
   useEffect(() => {
     (async () => {
       const token = params.get(ExtraParamsKey.OneTimeToken);
       const email = params.get(ExtraParamsKey.LoginHint);
       const errorMessage = params.get('errorMessage');
 
       if (errorMessage) {
-        setOneTimeTokenError(errorMessage);
+        navigate(
+          { pathname: `/${experience.routes.oneTimeToken}/error` },
+          { replace: true, state: { errorMessage } }
+        );
         return;
       }
 
       if (!token || !email) {
-        setOneTimeTokenError(true);
+        navigate(`/${experience.routes.oneTimeToken}/error`, { replace: true });
         return;
       }
 
-      /**
-       * Check if the user has agreed to the terms and privacy policy before navigating to the 3rd-party social sign-in page
-       * when the policy is set to `Manual`
-       */
+      // Terms gating (Manual policy): validate when required; termsValidation is idempotent
       if (agreeToTermsPolicy === AgreeToTermsPolicy.Manual && !(await termsValidation())) {
+        navigate(
+          { pathname: `/${experience.routes.oneTimeToken}/error` },
+          {
+            replace: true,
+            state: {
+              title: 'error.terms_acceptance_required',
+              message: 'error.terms_acceptance_required_description',
+            },
+          }
+        );
+        return;
+      }
+
+      // Ensure we only run this flow once per (token,email) pair
+      const runKey = `${token}|${email}`;
+      if (lastRunKeyRef.current === runKey) {
```

Comment:
Given the fact that the token or email should not be editable,  can we simply use an `isSubmitted` flag?

Reply:
I think yes, it should also work.

Reply:
Updated to `isSubmitted`